README.md
## VDPrequester

[![Join the chat at https://gitter.im/dthwaite/VDPrequester](https://badges.gitter.im/dthwaite/VDP.svg)](https://gitter.im/dthwaite/VDPrequester?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![npm](https://img.shields.io/npm/v/npm.svg)](https://www.npmjs.com/package/VDPrequester)
[![Build Status](https://secure.travis-ci.org/dthwaite/TPA.png)](http://travis-ci.org/dthwaite/VDPrequester)
[![Coverage Status](https://coveralls.io/repos/github/dthwaite/VDPrequester/badge.svg?branch=master)](https://coveralls.io/github/dthwaite/VDPrequester?branch=master)
[![bitHound Overall Score](https://www.bithound.io/github/dthwaite/VDPrequester/badges/score.svg)](https://www.bithound.io/github/dthwaite/VDPrequester)
[![bitHound Code](https://www.bithound.io/github/dthwaite/VDPrequester/badges/code.svg)](https://www.bithound.io/github/dthwaite/VDPrequester)

##### Requests jobs to be run in the cloud on volunteer computers

Available on [GitHub](https://github.com/dthwaite/VDPrequester), details on [JSDocs](http://dthwaite.github.io/docs/VDPrequester/1.0.0).

This is a very simple library that allows you to submit jobs into my experimental Volunteer Distributed Processing (VDP) network.
It is specific to this network and has no generic applicability.

##### Node.js:
To install it:
`npm install VDPrequester`

To see how to use it:
`npm docs`

To code with it:
```javascript
var vdprequester=new VDPrequester('ws://127.0.0.1:8080');

var square=21;
vdprequester.send(square,'squareroot',function(error,result) {
    'The square root of '+square+' is '+result.data;
});
```
##### Browser:
To install it:

* Download [vdprequester.min.js](https://github.com/dthwaite/VDPrequester/tree/master/lib/VDPrequester.min.js) from GitHub or use their CDN for my latest version: [v1.0.0](https://cdn.rawgit.com/dthwaite/VDPrequester/v1.0.0/lib/vdprequester.min.js)
* `vdprequester.min.js` is a UMD (Universal Module Definition) bundle with an export name of `VDPrequester`

To code with it:
```javascript
<script>
var vdprequester=new VDPrequester('ws://127.0.0.1:8080');

var square=21;
vdprequester.send(square,'squareroot',function(error,result) {
    'The square root of '+square+' is '+result.data;
});
</script>
```

##### Development:

Lint (to ensure no `eslint` issues):
`npm run lint`

Build minified version for browser into lib/vdprequester.min.js:
`npm run build-prod`

Build and run test for browser testing and debugging:
`npm test`

Build and run test with coverage:
`npm run test-coverage`

Review coverage:
`npm run view-coverage` (after `npm-run test-coverage`)

CHANGELOG.md
## VDPrequester

#### v1.0.0 26-04-16
* Initial release

test/test.js
/*global before,after,beforeEach,afterEach,describe, it, chai, sinon, VDPrequester*/
var assert=chai.assert;
//window.initMochaPhantomJS();
window.WebSocket=function() {

};
describe('VDP Requester', function() {
    before(function() {
        this.clock = sinon.useFakeTimers();
    });

    after(function() {
        this.clock.restore();
    });

    beforeEach(function() {
        this.ws=sinon.stub(window,'WebSocket');
        this.sendSpy=sinon.spy();
        WebSocket.prototype.send=this.sendSpy;
        WebSocket.prototype.close=function() {
            this.onclose();
        };
    });
    afterEach(function() {
        this.ws.restore();
    });
    it('Create a new link', function() {
        assert.equal(new VDPrequester('ws://127.0.0.1:8080').url, 'ws://127.0.0.1:8080');
        assert.throws(function() {
            new VDPrequester('http://127.0.0.1:8080');
        },'Invalid server address');
    });
    it('Attempt to send given no connection',function() {
        var data='data';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        this.ws.throws();
        vdprequester.send(library,data,callback);
        assert.equal(window.WebSocket.calledWithExactly(url,'R'+library),true);
        assert.equal(callback.calledWith('Connection to server refused'),true);
    });
    it('Successful first send',function() {
        var data='data';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        vdprequester.send(library,data,callback);
        assert.equal(window.WebSocket.calledWithExactly(url,'R'+library),true);
        assert.equal(callback.calledWith('Connection to server refused'),false);
        vdprequester.libraries[library][0].onopen();
        assert.equal(this.sendSpy.calledWithExactly(data),true);
    });
    it('Successful first send with server closing',function() {
        var data='data';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        vdprequester.send(library,data,callback);
        vdprequester.libraries[library][0].onopen();
        vdprequester.libraries[library][0].onclose();
        assert.equal(callback.calledWith('Server closed your connection'),true);
    });
    it('Initial failure to send followed by success',function() {
        var data='data';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        vdprequester.send(library,data,callback);
        var connection=vdprequester.libraries[library][0];
        connection.onopen();
        connection.onerror();
        connection.onclose();
        this.clock.tick(5000);
        connection=vdprequester.libraries[library][0];
        connection.onopen();
        assert.equal(this.sendSpy.calledTwice,true);
    });
    it('Repeated failure beyond 1 minute to send',function() {
        var data='data';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        vdprequester.send(library,data,callback);
        for (var i=0; i<12; i++) {
            var connection = vdprequester.libraries[library][0];
            connection.onopen();
            connection.onerror();
            connection.onclose();
            this.clock.tick(5000);
        }
        assert.equal(callback.callCount,0);
        connection = vdprequester.libraries[library][0];
        connection.onopen();
        connection.onerror();
        connection.onclose();
        this.clock.tick(5000);
        assert.equal(callback.calledWith('Server consistently unavailable'),true);
    });
    it('Two separate library requests',function() {
        var data1='data-1';
        var data2='data-2';
        var url='ws://127.0.0.1:8080';
        var library1='my-lib-1';
        var library2='my-lib-2';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();

        vdprequester.send(library1,data1,callback);
        vdprequester.libraries[library1][0].onopen();
        vdprequester.send(library2,data2,callback);
        vdprequester.libraries[library2][0].onopen();
        assert.equal(this.sendSpy.getCall(0).calledWithExactly(data1),true);
        assert.equal(this.sendSpy.getCall(1).calledWithExactly(data2),true);
    });
    it('Two same library requests',function() {
        var data1='data-1';
        var data2='data-2';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();

        vdprequester.send(library,data1,callback);
        vdprequester.libraries[library][0].onopen();
        vdprequester.send(library,data2,callback);
        vdprequester.libraries[library][1].onopen();
        assert.equal(this.sendSpy.getCall(0).calledWithExactly(data1),true);
        assert.equal(this.sendSpy.getCall(1).calledWithExactly(data2),true);
    });
    it('Successful first send with response',function() {
        var data='data';
        var result='result';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();

        vdprequester.send(library,data,callback);
        vdprequester.libraries[library][0].onopen();
        vdprequester.libraries[library][0].onmessage({data: result});
        assert.equal(this.sendSpy.calledWithExactly(data),true);
        assert.equal(callback.calledWithExactly(null,result),true);
    });
    it('Successful first send without callback',function() {
        var data='data';
        var result='result';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        vdprequester.send(library,data);
        vdprequester.libraries[library][0].onopen();
        vdprequester.libraries[library][0].onmessage({data: result});
        assert.equal(this.sendSpy.calledWithExactly(data),true);
    });
    it('Two same library requests with responses',function() {
        var data1='data-1';
        var data2='data-2';
        var result1='result-1';
        var result2='result-2';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        vdprequester.send(library,data1,callback);
        vdprequester.libraries[library][0].onopen();
        vdprequester.send(library,data2,callback);
        vdprequester.libraries[library][1].onopen();
        vdprequester.libraries[library][1].onmessage({data: result2});
        vdprequester.libraries[library][0].onmessage({data: result1});
        assert.equal(callback.getCall(0).calledWithExactly(null,result2),true);
        assert.equal(callback.getCall(1).calledWithExactly(null,result1),true);
    });
    it('Successive same library requests',function() {
        var data='data';
        var result='result';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var vdprequester=new VDPrequester(url);
        var callback=sinon.spy();
        vdprequester.send(library,data,callback);
        vdprequester.libraries[library][0].onopen();
        vdprequester.libraries[library][0].onmessage({data: result});
        assert.equal(callback.calledOnce,true);

        vdprequester.send(library,data,callback);
        vdprequester.libraries[library][0].onopen();
        vdprequester.libraries[library][0].onmessage({data: result});
        assert.equal(callback.calledTwice,true);

        assert.equal(this.ws.calledOnce,true);
    });
    it('Maximum connections',function() {
        var data='data';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var callback=sinon.spy();
        var vdprequester=new VDPrequester(url);
        for (var i=0; i<25; i++) {
            vdprequester.send(library, data, function() {});
            vdprequester.libraries[library][i].onopen();
            assert.equal(this.sendSpy.callCount,i+1);
        }
        vdprequester.send(library, data, callback);
        assert.equal(callback.calledWith('Maximum connection count reached'),true);
    });
    it('Disconnect',function() {
        var data='data';
        var result='result';
        var url='ws://127.0.0.1:8080';
        var library='my-lib';
        var callback1=sinon.spy();
        var callback2=sinon.spy();
        var callback3=sinon.spy();
        var callback4=sinon.spy();
        var vdprequester=new VDPrequester(url);
        vdprequester.send(library, data, callback1);
        vdprequester.send(library, data, callback2);
        vdprequester.libraries[library][1].onopen();
        vdprequester.send(library, data, callback3);
        vdprequester.libraries[library][2].onopen();
        vdprequester.libraries[library][2].onmessage({data: result});
        vdprequester.send(library, data, callback4);
        assert.equal(callback1.called,false);
        assert.equal(callback2.called,false);
        assert.equal(callback3.calledWithExactly(null,result),true);
        assert.equal(callback4.called,false);
        assert.equal(vdprequester.connectionCount(),3);
        vdprequester.disconnect();
        assert.equal(vdprequester.connectionCount(),0);
        assert.equal(callback1.called,false);
        assert.equal(callback2.called,false);
        assert.equal(callback3.calledOnce,true);
        assert.equal(callback4.called,false);
    });
});


test/vdprequester.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.VDPrequester = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
/**
 * @file Job request functionality for connection to Volunteer Distributed Processing
 * @author Dominic Thwaites
 * @copyright (c) 2016 Dominic Thwaites dominicthwaites@mac.com
 * @licence MIT
 * @module VDPrequester
 */
module.exports=(/** @lends module:VDPrequester*/function() {
    /**
     * The maximum number of connections that can be made - set to 25.
     * Note that connections for completed jobs are re-used and thus new connections are only made as jobs are requested in parallel.
     * @type {number}
     * @const
     */
    var MAX_CONNECTIONS=25;

    /**
     * Callback function template given when sending a job
     *
     * @callback requestCallback
     * @param {string} errorMessage Error message or null if no error
     * @param {(string|Blob|BufferArray)} resultData The data return from the job
     * @see VDPrequester#send
     */

    /**
     * Class from which any and all cloud jobs are sent to a particular server.
     *
     * Note that connections to the server are only made upon job submission. Thus a bad server
     * address given here will not be immediately apparent.
     *
     * @param {string} address The URL complete with port number that is a VDP server (ws:// ...)
     * @throws {Error} if the address is not a websocket address
     * @constructor
     */
    var VDPrequester=function(address) {
        if (typeof address=='string' && address.substr(0,5)=='ws://') {
            this.url = address;    // server URL
            this.libraries = {};   // VDPrequester holds a set of connections per library name
        } else throw new Error('Invalid server address');
    };

    /**
     * @returns {number} The number of connections open to the VDP server (max 25)
     */
    VDPrequester.prototype.connectionCount=function() {
        var count=0;
        for (var library in this.libraries) {
            count+=this.libraries[library].length;
        }
        return count;
    };

    /**
     * Sends a new job into the cloud.
     *
     * Generally, this is a fire and forget operation. There is no telling how long the job will take or when and who
     * will run the job. If you don't care (unlikely) then you need not provide a callback. But a callback is recommended
     * as it can tell about failures as well as the data results of the job.
     *
     * The library must be a valid name taken from the VDP repository.
     *
     * @param {string} library Name of the execution library to invoke
     * @param {string | Blob | bufferArray} data Input data to the library function
     * @param {requestCallback} [callback] Function called on completion of the job (passing error and result data as a parameter)
     */
    VDPrequester.prototype.send=function(library,data, callback) {
        if (typeof callback!='function') callback=function() {};
        this._send(new Request(library,data, callback,this));
    };

    /**
     * Cancels all outstanding job requests and closes all server connections.
     *
     * This can be used to clean up connections. Requests that have yet to complete are 'forgotten'
     * and thus no response will ever come back. However, those jobs are not cancelled and may or may not
     * complete in the cloud. This request object continues to be operative, however, allowing new requests
     * (and thus new connections) to be made to the server.
     */
    VDPrequester.prototype.disconnect=function() {
        for (var library in this.libraries) {
            var list=this.libraries[library].slice();
            for (var i=0; i<list.length; i++) {
                list[i].available=true;
                list[i].close();
            }
        }
    };

    // Sends one request to the server over a spare connection or a new one if necessary
    VDPrequester.prototype._send=function(request) {
        if (!(request.library in this.libraries)) this.libraries[request.library]=[];
        var list=this.libraries[request.library];
        for (var i = 0; i < list.length; i++) {
            if (list[i].available) {
                request.resetSocket(list[i]).send(request.data);
                return;
            }
        }
        if (this.connectionCount()>=MAX_CONNECTIONS) {
            request.callback('Maximum connection count reached');
        }
        try {
            list[i] = request.resetSocket(new WebSocket(this.url, 'R' + request.library));
            list[i].onopen = function() {
                list[i].send(request.data);
            };
        } catch (error) {
            request.callback('Connection to server refused');
        }
    };

    // Represents one job request
    var Request=function(library,data, callback,vdprequester) {
        this.library=library;
        this.data = data;
        this.vdprequester = vdprequester;
        this.callback = callback;
        this.errors=0;
    };

    // Sets the socket's callbacks in the context of this request
    Request.prototype.resetSocket=function(socket) {
        socket.available=false;
        var me=this;

        socket.onclose=function() {
            var list=me.vdprequester.libraries[me.library];
            // Remove this socket from our store
            list.splice(list.indexOf(socket),1);

            // Resend if we never got a response
            if (me.errors>0) {
                if (me.errors>12) me.callback('Server consistently unavailable');
                else {
                    setTimeout(function() {
                        me.vdprequester._send(me);
                    }, 5000);
                }
            } else {
                // Server purposefully closed this connection - meaning that
                // the message sent was not acceptable
                if (!socket.available) me.callback('Server closed your connection');
            }
        };
        socket.onmessage=function(event) {
            // flag as complete and send the data to the user
            socket.available = true;
            me.callback(null,event.data);
        };
        socket.onerror=function() {
            me.errors++;
        };
        return socket;
    };

    return VDPrequester;
})();

},{}]},{},[1])(1)
});

lib/vdprequester.js
/**
 * @file Job request functionality for connection to Volunteer Distributed Processing
 * @author Dominic Thwaites
 * @copyright (c) 2016 Dominic Thwaites dominicthwaites@mac.com
 * @licence MIT
 * @module VDPrequester
 */
module.exports=(/** @lends module:VDPrequester*/function() {
    /**
     * The maximum number of connections that can be made - set to 25.
     * Note that connections for completed jobs are re-used and thus new connections are only made as jobs are requested in parallel.
     * @type {number}
     * @const
     */
    var MAX_CONNECTIONS=25;

    /**
     * Callback function template given when sending a job
     *
     * @callback requestCallback
     * @param {string} errorMessage Error message or null if no error
     * @param {(string|Blob|BufferArray)} resultData The data return from the job
     * @see VDPrequester#send
     */

    /**
     * Class from which any and all cloud jobs are sent to a particular server.
     *
     * Note that connections to the server are only made upon job submission. Thus a bad server
     * address given here will not be immediately apparent.
     *
     * @param {string} address The URL complete with port number that is a VDP server (ws:// ...)
     * @throws {Error} if the address is not a websocket address
     * @constructor
     */
    var VDPrequester=function(address) {
        if (typeof address=='string' && address.substr(0,5)=='ws://') {
            this.url = address;    // server URL
            this.libraries = {};   // VDPrequester holds a set of connections per library name
        } else throw new Error('Invalid server address');
    };

    /**
     * @returns {number} The number of connections open to the VDP server (max 25)
     */
    VDPrequester.prototype.connectionCount=function() {
        var count=0;
        for (var library in this.libraries) {
            count+=this.libraries[library].length;
        }
        return count;
    };

    /**
     * Sends a new job into the cloud.
     *
     * Generally, this is a fire and forget operation. There is no telling how long the job will take or when and who
     * will run the job. If you don't care (unlikely) then you need not provide a callback. But a callback is recommended
     * as it can tell about failures as well as the data results of the job.
     *
     * The library must be a valid name taken from the VDP repository.
     *
     * @param {string} library Name of the execution library to invoke
     * @param {string | Blob | bufferArray} data Input data to the library function
     * @param {requestCallback} [callback] Function called on completion of the job (passing error and result data as a parameter)
     */
    VDPrequester.prototype.send=function(library,data, callback) {
        if (typeof callback!='function') callback=function() {};
        this._send(new Request(library,data, callback,this));
    };

    /**
     * Cancels all outstanding job requests and closes all server connections.
     *
     * This can be used to clean up connections. Requests that have yet to complete are 'forgotten'
     * and thus no response will ever come back. However, those jobs are not cancelled and may or may not
     * complete in the cloud. This request object continues to be operative, however, allowing new requests
     * (and thus new connections) to be made to the server.
     */
    VDPrequester.prototype.disconnect=function() {
        for (var library in this.libraries) {
            var list=this.libraries[library].slice();
            for (var i=0; i<list.length; i++) {
                list[i].available=true;
                list[i].close();
            }
        }
    };

    // Sends one request to the server over a spare connection or a new one if necessary
    VDPrequester.prototype._send=function(request) {
        if (!(request.library in this.libraries)) this.libraries[request.library]=[];
        var list=this.libraries[request.library];
        for (var i = 0; i < list.length; i++) {
            if (list[i].available) {
                request.resetSocket(list[i]).send(request.data);
                return;
            }
        }
        if (this.connectionCount()>=MAX_CONNECTIONS) {
            request.callback('Maximum connection count reached');
        }
        try {
            list[i] = request.resetSocket(new WebSocket(this.url, 'R' + request.library));
            list[i].onopen = function() {
                list[i].send(request.data);
            };
        } catch (error) {
            request.callback('Connection to server refused');
        }
    };

    // Represents one job request
    var Request=function(library,data, callback,vdprequester) {
        this.library=library;
        this.data = data;
        this.vdprequester = vdprequester;
        this.callback = callback;
        this.errors=0;
    };

    // Sets the socket's callbacks in the context of this request
    Request.prototype.resetSocket=function(socket) {
        socket.available=false;
        var me=this;

        socket.onclose=function() {
            var list=me.vdprequester.libraries[me.library];
            // Remove this socket from our store
            list.splice(list.indexOf(socket),1);

            // Resend if we never got a response
            if (me.errors>0) {
                if (me.errors>12) me.callback('Server consistently unavailable');
                else {
                    setTimeout(function() {
                        me.vdprequester._send(me);
                    }, 5000);
                }
            } else {
                // Server purposefully closed this connection - meaning that
                // the message sent was not acceptable
                if (!socket.available) me.callback('Server closed your connection');
            }
        };
        socket.onmessage=function(event) {
            // flag as complete and send the data to the user
            socket.available = true;
            me.callback(null,event.data);
        };
        socket.onerror=function() {
            me.errors++;
        };
        return socket;
    };

    return VDPrequester;
})();
